src/chart/utils/linkedList.utils.ts

```diff
@@ -0,0 +1,135 @@
+export class ListNode<T> {
+	public data: T;
+	public next: ListNode<T> | null;
+
+	constructor(data: T) {
+		this.data = data;
+		this.next = null;
+	}
+}
+/**
+ * Implementation of Linked list data structure.
+ * @param head
+ * @param epsilon
+ */
+export class LinkedList<T> {
+	private head: ListNode<T> | null = null;
+	private length = 0;
+
+	constructor(head?: ListNode<T>) {
+		this.head = head ?? null;
+	}
+
+	public insertAtEnd(data: T): ListNode<T> {
+		const node = new ListNode(data);
+		let current: ListNode<T>;
+
+		if (this.head === null) {
+			this.head = node;
+		} else {
+			current = this.head;
+			while (current.next) {
+				current = current.next;
+			}
+			current.next = node;
+		}
+		this.length++;
+		return node;
+	}
+
+	public insertAt(position: number, data: T) {
+		if (position > -1 && position < this.length && this.head) {
+			let current = this.head;
+			let index = 0;
+			let previous = current;
+			const node = new ListNode(data);
+
+			if (position === 0) {
+				node.next = current;
+				this.head = node;
+			} else {
+				while (index++ < position && current.next) {
+					previous = current;
+					current = current.next;
+				}
+				node.next = current;
+				previous.next = node;
+			}
+			this.length++;
+			return current;
+		} else {
+			return null;
+		}
+	}
+
+	public removeAt(position: number): ListNode<T> | null {
+		if (position > -1 && position < this.length && this.head) {
+			let current = this.head;
+			let previous: ListNode<T> = current;
+			let index = 0;
+
+			if (position === 0) {
+				this.head = current.next;
+			} else {
+				while (index++ < position && current.next) {
+					previous = current;
+					current = current.next;
+				}
+				previous.next = current.next;
+			}
+			this.length--;
+			return current;
+		} else {
+			return null;
+		}
+	}
+
+	public getNodePosition(node: ListNode<T>) {
+		let index = 0;
+		let tempNode = this._head;
+
+		while (node) {
+			if (tempNode?.data === node.data) {
+				return index;
+			}
+			tempNode = tempNode && tempNode.next;
+			index++;
+		}
+
+		return -1;
+	}
+
+	public size(): number {
+		return this.length;
+	}
+
+	get _head() {
+		return this.head;
```

Comment:
usually internal fields stored like this `_head` and publicly availabe properties are like this `head`